lib/galaxy/jobs/command_factory.py

```diff
@@ -229,24 +233,37 @@ def __init__(self, initial_command=u''):
         # the last thing that happens is an exit with return code.
         self.return_code_captured = False
 
-    def prepend_command(self, command):
+    def prepend_command(self, command, sep=";"):
         if command:
-            self.commands = u"%s; %s" % (command,
+            self.commands = u"%s%s %s" % (command,
+                                         sep,
                                          self.commands)
         return self
 
     def prepend_commands(self, commands):
         return self.prepend_command(u"; ".join(c for c in commands if c))
 
-    def append_command(self, command):
+    def append_command(self, command, sep=';'):
         if command:
-            self.commands = u"%s; %s" % (self.commands,
-                                         command)
+            self.commands = u"%s%s %s" % (self.commands,
+                                          sep,
+                                          command)
         return self
 
     def append_commands(self, commands):
         self.append_command(u"; ".join(c for c in commands if c))
 
+    def capture_stdout_stderr(self, stdout_file, stderr_file):
+        self.prepend_command("""out="${TMPDIR:-/tmp}/out.$$" err="${TMPDIR:-/tmp}/err.$$"
+mkfifo "$out" "$err"
+trap 'rm "$out" "$err"' EXIT
+tee -a stdout.log < "$out" &
+tee -a stderr.log < "$err" >&2 &""",
```

Comment:
mkfifo and tee are on alpine:3.8, but not trap :-(. We need a replacement for it or handle it somehow. I tried with `docker run -it alpine:3.8 sh` and checked inside, otherwise this would make Galaxy incompatible with tools built on top of alpine.

Reply:
well I searched by `which` but I think trap could be a shell built-in? Then my comment wouldn't apply of course....

Reply:
Are you sure about that? This script works for me on `alpine:3.8`
```
#!/bin/sh

out="${TMPDIR:-/tmp}/out.$$" err="${TMPDIR:-/tmp}/err.$$"
mkfifo "$out" "$err"
trap 'rm "$out" "$err"' EXIT
tee -a stdout.log < "$out" &
tee -a stderr.log < "$err" >&2 &
bwa >"$out" 2>"$err"
echo $? > exit_code
```

also trap seems to be there: `command -v trap`

Reply:
`trap` should be ok.  It's a shell command compatible with both `sh` and `bash`

```
17:05 [pireddu@slynx ~] docker run -it alpine:3.8 sh
/ # trap 'echo "bye bye"' EXIT
/ # bye bye
```

Reply:
FYI you can use `command -V trap` to find out if it's a builtin, alias or normal executable:
```
$ command -V trap
trap is a shell builtin
```

Reply:
Scienziato! :bowing_man: :smile: